### DISCUSSION TITLE
Engineering: private and protected properties are being mangled now

### DISCUSSION THREAD

**Discussion by jrieken** [microsoft/vscode-discussions#257] [Announcements]
This is a heads-up that we have enabled mangling of private and protected class fields (properties and methods) across all our bundles. We are doing this to save _bandwidth_ and _startup time_. Mangling allowed us reduce the size of our renderer JS bundles by ~13% (-1.5MB). This yields in ~5% faster code loading. Similar numbers apply to the main and extension host bundles. The sample below illustrates what that means. 

```ts
class Animal {
  private name: number;
  public toString() {
    return `Animal: ${this.name}`
  } 
}

// ⬇️ after mangling

class Animal {
  private a: number;
  public toString() {
    return `Animal: ${this.a}`
  } 
}
```

🔈 *Note* that this should be of **no concern** for every extension that stick to the API as defined in `vscode.d.ts`. Meaning the vast majority of extensions aren't affected by this and will benefit from our improved startup times.

However, extensions that reach into the internals of how the API is implemented or extensions that patch VS Code itself will be impacted because all names have been changed now. Extending by reverse engineering or patching is discouraged and we invite every author of such extensions to get in touch with us so that we identify gaps and shortcomings in our extensibility story with the goal of making our actual API more rich. 

Happy Coding.

**Comment by r3m0t** [comment]
My codebase only has `as any` cast in code relating to `vscode-languageserver-node` so I hope no mangling will be applied in that NPM package.

**Reply by jrieken** [reply]
> I hope no mangling will be applied in that NPM package

Currently not but it what you do can break anytime, e.g by adopting property mangling, by adopting native private fields, or by any internal code change.

**Comment by RandomFractals** [comment]
does it affect webview extensions? or not this time?

**Comment by asumagic** [comment]
> However, extensions that reach into the internals of how the API is implemented or extensions that patch VS Code itself will be impacted because all names have been changed now. Extending by reverse engineering or patching is discouraged and we invite every author of such extensions to get in touch with us so that we identify gaps and shortcomings in our extensibility story with the goal of making our actual API more rich.

I am not the developer of such an extension, but "Customize UI" is one I was using, and it breaking got plenty of people disappointed as you've probably noticed.  
Personally, the settings it provided were exactly what I was trying to do to customize VSC, and the extension was essentially create to respond to years-old feature requests that have not been solved yet -- which, to be clear, _is perfectly fine_. I imagine there are good reasons for some of these.

That being said, to my limited understanding, wouldn't extending the extension API to enable what Customize UI was doing about as much work as adding in these specific customization features in the first place, if not more? It seems like it'd be difficult to implement this without setting too many things in stone while ensuring good extension forward compatibility.  
Considering this, unless these customization features jumped in priority due to the breakage of that extension, I fear this tradeoff is set to leave a hole in VSC customization capabilities (and thus in usability for a number of users) reopened for months or years.  
It's not the end of the world but it makes me wonder if the tradeoff is worth it just yet.

**Reply by bisho** [reply]
At the very least, work with CustomizeUI author to allow-list the private properties it relies on, and avoid mangling those!

**Comment by SamJakob** [comment]
> Extending by reverse engineering or patching is discouraged and we invite every author of such extensions to get in touch with us so that we identify gaps and shortcomings in our extensibility story with the goal of making our actual API more rich.

The only reason this has been done by any of the extension in the first place is to accommodate features or issues that the VS code team cannot or will not address (apparently, as established over a period of years) and users who have done so have accepted the ramifications of potential incompatibility, glitchiness or spending time of their own debugging issues.

> We are doing this to save bandwidth and startup time. Mangling allowed us reduce the size of our renderer JS bundles by ~13% (-1.5MB). This yields in ~5% faster code loading.

Extensions like Customize UI have allowed developers to make their own subjective changes, that probably wouldn't be of concern to anyone else but held to great personal importance. I personally have used the extension to patch VS code across many versions before this fundamental change made it impossible - **causing such a change for 5% increase in load time and 1.5 MiB decrease in code size has a heavy cost associated in terms of lost productivity, bad-will/bad faith, and lost functionality with almost no tangible difference to anyone whatsoever.**

> Extending by reverse engineering or patching is discouraged and we invite every author of such extensions to get in touch with us so that we identify gaps and shortcomings in our extensibility story with the goal of making our actual API more rich.

As mentioned before (but I felt this was worth re-iterating) the only reasons these projects have been started is because after a few years of having issues open or closed without resolution, power users have taken it upon themselves to develop niche features for themselves:
https://github.com/microsoft/vscode/issues/519#issuecomment-1184316289
https://github.com/microsoft/vscode/issues/118692

---

In essence, this change seems rather futile and seemingly serves only to remove functionality that despite being incredibly useful and valued to many power users likely will not re-appear in the official API either due to time or resource constraints, being out-of-scope or irrelevant to the majority of users, or for political reasons.

I should expect that any user either 'waits for VS code to open' or they don't - these extremely marginal gains in terms of load time would be unlikely to have any impact and when a browser uses 1 GiB of RAM for a few tabs, I'm sure a few megabytes for the added functionality wouldn't be missed either.

**Reply by chakravarthi-bb** [reply]
Can we ship this mangled version only for the browsers??

**Reply by SamJakob** [reply]
> Can we ship this mangled version only for the browsers??

Off the top of my head, this suggestion (to ship this change only for browsers) would make **a lot more** sense so I agree with that approach. At least then, the benefit of the changes is somewhat more meaningful. But this **absolutely should not** be done to the detriment of this functionality on the desktop as for the above reasons the loss absolutely does not outweigh the benefit.

**Comment by NuckChorris** [comment]
This provides a tiny startup performance gain and tinier bundle size improvement (less than 0.5%!) at the cost of serious good will.

You should absolutely revert this micro-optimization to fix monkey-patch in a 1.74.2 patch release and _then_ figure out what options you need to provide to replace it. Maybe start with some of the stuff you closed over the past 7 years like Microsoft/vscode#38555 and microsoft/vscode#148521 microsoft/vscode#82996 and microsoft/vscode#66639. Or maybe one that's sat open for 6 years like microsoft/vscode#12377.

People have been begging for you to address these issues for years, and gave up and started hacking them in through monkey-patch. I, for one, am not upgrading my VS Code until this is addressed.

**Reply by SamJakob** [reply]
This.

The decision to deprive users of the features they've been fruitlessly requesting for years - _and then written/resolved themselves_ - for an almost unnoticeable micro-optimization is positively mind boggling. The audacity, to then say that users need to do the process over again as though there's any reason this time should be different is simply astounding. Not to mention that, at best, in terms of functionality it will probably be more or less the same as it was before this change once all is said and done (seeing as the end goal is to just re-add the functionality after determining whether it's worth the VS Code team's time).

There's such a lack of sane (or practical) rationale in this decision that it's really, really, hard to take these actions as being done in good faith, or rather, without bad faith.

It's just irritating to feel so unheard, as developers, by developers writing a tool for developers _whilst_ being asking for feedback that in all probability will be ignored.

**Comment by chakravarthi-bb** [comment]
Thanks, Microsoft!! for listening to user feedback. It was good while it lasted. 

https://en.wikipedia.org/wiki/Embrace,_extend,_and_extinguish

**Comment by veryjos** [comment]
Is the impact _actually_ 5% on startup? That seems a little high to me.

I understand poking at internals is problematic for a dev team, but nobody expected support for their patches-- they just didn't expect the capability to be removed altogether 😞 

I patched basically anything that annoyed me about VSCode-- and now it's way less fun to use, at the very least.

Downgrading for now, but really hoping this is addressed.. somehow.

**Comment by luan** [comment]
I have been silently watching this issue to avoid piling up on an already well established argument. But now I'm posting to make sure other people that are doing the same know that there are many of us here. The proposed long term solutions would be great, but we don't want to wait for long-term to get functionality we had last week back.

I get the argument you made about not wanting to encourage monkey-patching. But I think that's just ill advised, from a community standpoint, enabling any and all tinkering should in fact be encouraged. The core team should then keep and eye and open communication with the tinkerers to find the gaps. If there's no tinkering then we will flock elsewhere. I had very recently migrated from Vim because I had found the Customizse UI and Monkey Patch extensions that allowed me to make VSCode truly the editor I wanted to use.

The issues it helps address aren't new, we haven't been hiding behind the hacking, there are many issues on the VSCode repo about the many creature comforts we've wanted.

This is, without a doubt, a breaking change. It doesn't matter if the API wasn't meant to be public, it is an API that was being widely used, and for all intents and purposes, was public, even if by accident. Therefore, this is absolutely **not** how you make a breaking change. First you need to provide a migration path, then you deprecated something and only then you remove the API.

There is no other way to address this then reverting the PR that introduced the breaking change and only then talking about the long-term change.

**Comment by zawilliams** [comment]
Just posting in here because nothing seems to be happening in a positive way to address the issue with Customize UI. I'm also one of the "silent" followers of this issue. Just want to make my voice known.

Mangling doesn't make sense for the desktop app. No one cares if the bundle size is bigger. We need our customizations back that make VS Code our preferred editor. This change should have been made in a major release and not in a regular update.

**Reply by AubreyF** [reply]
^^ This. Not being able to turn off window title bars on MacOS is a dealbreaker for me. Like, seriously, installing-fleet-now-caliber dealbreaker.

Related: 
- https://github.com/iocave/customize-ui/issues/156
- https://github.com/microsoft/vscode/issues/12377
- https://github.com/microsoft/vscode/pull/166126

**Comment by panoskj** [comment]
I haven't researched this issue much yet (I just updated VSCode and noticed it), but isn't it a common practice to maintain a file mapping mangled names to original names, to ease debugging for example? 

Since the VSCode's code is open-source and there is nothing to hide, why not provide a mapping for each version and let the extensions' developers figure out the rest?

I mean, people have been successfully modding minecraft for years, even though its code is obfuscated. They even made and maintained the mappings themselves. So, if this is an attempt to prevent modding VSCode's internals, it really sounds like a pointless endeavor to me. 

On the other hand, if this isn't an attempt to prevent such modifications, then there is no explanation why you did not communicate with the authors of the affected extensions first.

**Reply by gjsjohnmurray** [reply]
@panoskj wrote:

> there is no explanation why you did not communicate with the authors of the affected extensions first

Quoting from the message with which @jrieken started this thread:

> we invite every author of such extensions to get in touch with us

Only the extension author is in a position to know whether or not they will be affected.

**Reply by panoskj** [reply]
@gjsjohnmurray  Sorry I was not very clear, but I had "Customize UI" extension in mind and I meant communicating with them to find a solution before releasing such an update. If there are any other popular extensions that broke, communicating with their developers instead would also work out the same.

So, "Customize UI" says in its description: "Customize UI relies on the Monkey Patch Extension to inject custom javascript in VSCode". My point is, it should be relatively easy/obvious to find this extension will break, that is, if you had looked for the most popular extensions that will break. Moreover there is an unresolved "think about generating source maps and check if TS ingests source maps" task over here: https://github.com/microsoft/vscode/pull/166126.

In my opinion, at this point it simply didn't make any sense to obfuscate VSCode without providing the mappings first. It looks like a rushed release.

**Comment by ghost** [comment]
I lost my love "Customize UI". 
And i along with 100k+ users blame you for this.

You. Yes, YOU made us downgrade vscode. Soon many will be leaving.

Outrageous. Open source? Sh*t

**Comment by luan** [comment]
@jrieken There's a lot of attempts here to reach out to you guys to start conversations but in several weeks we've seen nothing that even remotely resembles an olive branch. I realize that some folks have been a bit heated in how they brought issues up, which can be counterproductive. But not engaging is not helping the situation.

**Comment by bitboxx** [comment]
@jrieken / @isidorn Could you please post an update on this??

I'm a Monkey Patch + Customize UI user to set UI custom fonts and CSS. This update ruined it.

**Comment by bisho** [comment]
I downgraded and disabled updates. :(

**Comment by shellscape** [comment]
I don't buy the reasoning for a second, given the disdain that MS employees have expressed in issues for any extension that modifies the UX beyond what they want to control. This was a malicious act explicitly targeting extensions which work around the gate-keeping by this team. This is yet another signal that the Microsoft team does not give any care to the user base, only what serves Microsoft, and that they will gatekeep the UX at all costs.

Microsoft are not good stewards of open source, and this further proves as much.

**Comment by thefakeplace** [comment]
Can Monkey Patch generate a translation table? In another thread someone mentioned MS could generate one, but we should be able to as well.

It is an unnecessary complication, sure-- but it should be possible, no? I don't think MS will prioritize fixing this.

**Comment by isidorn** [comment]
We are going to lock this issue because feedback isn’t constructive anymore.

To folks with constructive energy: we do have source map information for mangled properties and anyone can use those to reverse map the non-mangled names. Fundamentally, patching VS Code is still possible it just got harder because names are shortened (for the reasons outlined above). Generally, extension relying on patching internals are tolerated but not supported. Doing so would put us into a position where absolutely no change is possible anymore. Future engineering work, like the upcoming AMD to ESM migration, can cause a similar swirl and we will not pause them.

We understand the frustration of your specific feature request not being prioritised or API being developed slowly but that’s simply the way it is. Screaming louder or insulting us won’t help in the process.

Extensions that access private and protected fields are not using API, they are using workarounds to patch the VS Code experience. The author of the Customise UI extension confirmed that [here](https://github.com/iocave/customize-ui/issues/156#issuecomment-1325708129).

Thanks for understanding
